### PR TITLE
Improve profile API

### PR DESCRIPTION
**Pull request by kaituo**
*Issue #, if available:*

*Description of changes:*
This PR did various things to improve profile API:
First, the PR fixed the hang issue. Previously, when users run _profile/state or _profile on the multi-entity detector, the request hangs. The problem is due to incorrect maxResponseCount passed to MultiResponsesDelegateActionListener.
Second, the PR fixes the multi-entity detector's wrong state issue. Previously, we can show the init state after an anomaly has shown up. We may have the problem because we read the most active entity's init progress in the cache for a detector's init_progress. But the entity already produced anomaly has been evicted out of the cache. This PR fixes the issue by double-checking the result index's non-zero RCF score for a multi-entity detector before reporting the init state. If there is any non-zero RCF score, we say running state instead of the initing state.
Third, this PR adds more information to the entity level profile, including last_active_timestamp, last_sample_timestamp, init_progress, model, and state.
Fourth, this PR adds models and total_size_in_bytes to the multi-entity detector level profile.

This PR also fixes various "fail to return" issues in the rest API related transport action. We didn't return after sending channel responses. Later, when we use the channel to send back responses again, we get " java.lang.IllegalStateException: Channel is already closed."

Testing done:
1. manual testing passes.
2. added unit tests

After the change, we have the following output for multi-entity detectors:

```
http://localhost:9200/_opendistro/_anomaly_detection/detectors/T4c3dXUBj-2IZN7itix_/_profile?_all=true&pretty

{
	"state": "RUNNING",
	"models": [{
			"model_id": "T4c3dXUBj-2IZN7itix__entity_app_4",
			"model_size_in_bytes": 712480,
			"node_id": "g6pmr547QR-CfpEvO67M4g"
		},
		{
			"model_id": "T4c3dXUBj-2IZN7itix__entity_app_5",
			"model_size_in_bytes": 712480,
			"node_id": "g6pmr547QR-CfpEvO67M4g"
		},
		{
			"model_id": "T4c3dXUBj-2IZN7itix__entity_app_6",
			"model_size_in_bytes": 712480,
			"node_id": "g6pmr547QR-CfpEvO67M4g"
		},
		{
			"model_id": "T4c3dXUBj-2IZN7itix__entity_app_0",
			"model_size_in_bytes": 712480,
			"node_id": "g6pmr547QR-CfpEvO67M4g"
		},
		{
			"model_id": "T4c3dXUBj-2IZN7itix__entity_app_1",
			"model_size_in_bytes": 712480,
			"node_id": "g6pmr547QR-CfpEvO67M4g"
		},
		{
			"model_id": "T4c3dXUBj-2IZN7itix__entity_app_2",
			"model_size_in_bytes": 712480,
			"node_id": "g6pmr547QR-CfpEvO67M4g"
		},
		{
			"model_id": "T4c3dXUBj-2IZN7itix__entity_app_3",
			"model_size_in_bytes": 712480,
			"node_id": "g6pmr547QR-CfpEvO67M4g"
		}
	],
	"total_size_in_bytes": 4987360,
	"init_progress": {
		"percentage": "100%"
	},
	"total_entities": 7,
	"active_entities": 7
}

http://localhost:9200/_opendistro/_anomaly_detection/detectors/T4c3dXUBj-2IZN7itix_/_profile?_all=true&entity=app_6
{
    "category_field": "service",
    "value": "app_6",
    "is_active": true,
    "last_active_timestamp": 1604026394879,
    "last_sample_timestamp": 1604026394879,
    "init_progress": {
        "percentage": "100%"
    },
    "model": {
        "model_id": "TFUdd3UBBwIAGQeRh5IS_entity_app_6",
        "model_size_in_bytes": 712480,
        "node_id": "MQ-bTBW3Q2uU_2zX3pyEQg"
    },
    "state": "RUNNING"
}

http://localhost:9200/_opendistro/_anomaly_detection/detectors/T4c3dXUBj-2IZN7itix_/_profile/entity_info,init_progress?entity=app_0

{
    "category_field": "service",
    "value": "app_0",
    "is_active": true,
    "last_active_timestamp": 1604020861321,
    "last_sample_timestamp": 1604020861321,
    "init_progress": {
        "percentage": "100%"
    }
}
```

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/main/java/com/amazon/opendistroforelasticsearch/ad/AnomalyDetectorProfileRunner.java [hunk 1]
```diff
@@ -70,82 +71,112 @@
...
+                        }
+                        if (profilesToCollect.contains(DetectorProfileName.COORDINATING_NODE)
+                            || profilesToCollect.contains(DetectorProfileName.SHINGLE_SIZE)
+                            || profilesToCollect.contains(DetectorProfileName.TOTAL_SIZE_IN_BYTES)
+                            || profilesToCollect.contains(DetectorProfileName.MODELS)
+                            || profilesToCollect.contains(DetectorProfileName.ACTIVE_ENTITIES)
+                            || profilesToCollect.contains(DetectorProfileName.INIT_PROGRESS)
+                            || profilesToCollect.contains(DetectorProfileName.STATE)) {
+                            totalResponsesToWait++;
+                        }
```
Comment (yizheliu-amazon) [inline src/main/java/com/amazon/opendistroforelasticsearch/ad/AnomalyDetectorProfileRunner.java:142]: looks like we can combine these 2 `if` into single one

File: src/main/java/com/amazon/opendistroforelasticsearch/ad/caching/EntityCache.java [hunk 2]
```diff
@@ -81,4 +83,19 @@
...
+     * is lastly accessed (get/put).  If the entity's state is inactive in the cache,
+     * the value indicates when the cache state is created or when the entity is evicted
+     * from active entity cache.
+     *
+     * @param detectorId The Id of the detector that an entity belongs to
+     * @param entityModelId Entity's Model Id
+     * @return if the entity is in the cache, return the timestamp in epoch
+     * milliseconds when the entity's state is lastly used.  Otherwise, return -1.
+     */
+    long getLastActiveMs(String detectorId, String entityModelId);
```
Comment (yizheliu-amazon) [inline src/main/java/com/amazon/opendistroforelasticsearch/ad/caching/EntityCache.java:100]: just name it `getLastActiveModels`

**Review by yizheliu-amazon — APPROVED** [review]
thanks for the change.

File: src/main/java/com/amazon/opendistroforelasticsearch/ad/AnomalyDetectorProfileRunner.java (hunk 1 above)
Comment (kaituo) [reply src/main/java/com/amazon/opendistroforelasticsearch/ad/AnomalyDetectorProfileRunner.java:142]: I separate them on purpose.  Each group will cost MultiResponsesDelegateActionListener one response.

File: src/main/java/com/amazon/opendistroforelasticsearch/ad/caching/EntityCache.java (hunk 2 above)
Comment (kaituo) [reply src/main/java/com/amazon/opendistroforelasticsearch/ad/caching/EntityCache.java:100]: For "Ms", I meant milliseconds.  Please see https://en.wikipedia.org/wiki/Millisecond.

File: src/main/java/com/amazon/opendistroforelasticsearch/ad/AnomalyDetectorProfileRunner.java [hunk 3]
```diff
@@ -409,76 +358,126 @@ private void profileStateRelated(
...
+        AnomalyDetector detector,
+        long totalUpdates,
+        MultiResponsesDelegateActionListener<DetectorProfile> listener
+    ) {
+        return ActionListener.wrap(searchResponse -> {
+            SearchHits hits = searchResponse.getHits();
+            if (hits.getTotalHits().value == 0L) {
+                processInitResponse(detector, profilesToCollect, totalUpdates, false, profileBuilder, listener);
+            } else {
+                createRunningStateAndInitProgress(profilesToCollect, profileBuilder);
```
Comment (ylwu-amzn) [inline src/main/java/com/amazon/opendistroforelasticsearch/ad/AnomalyDetectorProfileRunner.java:465]: If a running detector stopped, then restart but not pass initialization yet. We can find anomaly results with anomaly score > 0 as the detector was running before. We can't tell the detector is at running status exactly for this case.

Comment (kaituo) [reply src/main/java/com/amazon/opendistroforelasticsearch/ad/AnomalyDetectorProfileRunner.java:465]: I am searching records older than the job's enabled time.  Does that cover the issue you mentioned?

Comment (ylwu-amzn) [reply src/main/java/com/amazon/opendistroforelasticsearch/ad/AnomalyDetectorProfileRunner.java:465]: Cool, make sense.

**Review by ylwu-amzn — APPROVED** [review]

**Pull request merged**